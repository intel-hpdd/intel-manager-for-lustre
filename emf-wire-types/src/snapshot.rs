// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

//! Data structures for communicating with the agent regarding lustre snapshots.

use crate::{
    db::{Id, TableName},
    graphql_duration::GraphQLDuration,
};
use chrono::{offset::Utc, DateTime};
use std::str::FromStr;
#[cfg(feature = "cli")]
use structopt::StructOpt;

#[derive(serde::Deserialize, serde::Serialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to list snapshots
pub struct List {
    /// Target name
    pub target: String,
    /// Name of the snapshot to list
    pub name: Option<String>,
}

#[derive(serde::Deserialize, serde::Serialize, Clone, PartialEq, Debug)]
#[cfg_attr(feature = "graphql", derive(juniper::GraphQLObject))]
/// Snapshots description
pub struct Snapshot {
    pub filesystem_name: String,
    pub snapshot_name: String,
    /// Snapshot filesystem id (random string)
    pub snapshot_fsname: String,
    pub modify_time: DateTime<Utc>,
    pub create_time: DateTime<Utc>,
    pub mounted: bool,
    /// Optional comment for the snapshot
    pub comment: Option<String>,
}

#[derive(serde::Deserialize, serde::Serialize, Clone, PartialEq, Debug)]
pub struct SnapshotRecord {
    pub id: i32,
    pub filesystem_name: String,
    pub snapshot_name: String,
    pub modify_time: DateTime<Utc>,
    pub create_time: DateTime<Utc>,
    pub snapshot_fsname: String,
    pub mounted: bool,
    pub comment: Option<String>,
}

impl Id for SnapshotRecord {
    fn id(&self) -> i32 {
        self.id
    }
}

pub const SNAPSHOT_TABLE_NAME: TableName = TableName("snapshot");

#[cfg_attr(feature = "graphql", derive(juniper::GraphQLObject))]
#[derive(serde::Deserialize, serde::Serialize, Clone, PartialEq, Debug)]
/// A Snapshot interval
pub struct SnapshotInterval {
    /// The configuration id
    pub id: i32,
    /// The filesystem name
    pub filesystem_name: String,
    /// Use a write barrier
    pub use_barrier: bool,
    /// The interval configuration
    pub interval: GraphQLDuration,
    // Last known run
    pub last_run: Option<DateTime<Utc>>,
}

impl Id for SnapshotInterval {
    fn id(&self) -> i32 {
        self.id
    }
}

pub const SNAPSHOT_INTERVAL_TABLE_NAME: TableName = TableName("snapshot_interval");

#[cfg_attr(feature = "graphql", derive(juniper::GraphQLObject))]
#[derive(serde::Deserialize, serde::Serialize, Clone, PartialEq, Debug)]
pub struct SnapshotRetention {
    pub id: i32,
    pub filesystem_name: String,
    /// Amount or percent of free space to reserve
    pub reserve_value: i32,
    pub reserve_unit: ReserveUnit,
    /// Minimum number of snapshots to keep
    pub keep_num: i32,
    pub last_run: Option<DateTime<Utc>>,
}

impl Id for SnapshotRetention {
    fn id(&self) -> i32 {
        self.id
    }
}

pub const SNAPSHOT_RETENTION_TABLE_NAME: TableName = TableName("snapshot_retention");

#[cfg_attr(feature = "graphql", derive(juniper::GraphQLEnum))]
#[cfg_attr(feature = "postgres-interop", derive(sqlx::Type))]
#[cfg_attr(feature = "postgres-interop", sqlx(rename = "snapshot_reserve_unit"))]
#[cfg_attr(feature = "postgres-interop", sqlx(rename_all = "lowercase"))]
#[derive(serde::Deserialize, serde::Serialize, Clone, Copy, PartialEq, Debug)]
#[serde(rename_all = "lowercase")]
pub enum ReserveUnit {
    #[cfg_attr(feature = "graphql", graphql(name = "percent"))]
    Percent,
    #[cfg_attr(feature = "graphql", graphql(name = "gibibytes"))]
    Gibibytes,
    #[cfg_attr(feature = "graphql", graphql(name = "tebibytes"))]
    Tebibytes,
}

impl FromStr for ReserveUnit {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "%" | "percent" => Ok(Self::Percent),
            "gib" | "g" | "gibibytes" => Ok(Self::Gibibytes),
            "tib" | "t" | "tebibytes" => Ok(Self::Tebibytes),
            x => Err(format!("Unexpected '{}'", x)),
        }
    }
}

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to create a snapshot
pub struct Create {
    /// Filesystem name
    pub fsname: String,
    /// Snapshot name
    pub name: String,
    /// Set write barrier before creating snapshot
    #[cfg_attr(feature = "cli", structopt(short = "b", long = "use_barrier"))]
    pub use_barrier: bool,
    /// Optional comment for the snapshot
    #[cfg_attr(feature = "cli", structopt(short = "c", long = "comment"))]
    pub comment: Option<String>,
}

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to destroy the snapshot
pub struct Destroy {
    /// Filesystem name
    pub fsname: String,
    /// Name of the snapshot to destroy
    pub name: String,

    /// Destroy the snapshot by force
    #[cfg_attr(feature = "cli", structopt(short = "f", long = "force"))]
    pub force: bool,
}

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to mount the snapshot
pub struct Mount {
    /// Filesystem name
    pub fsname: String,
    /// Snapshot name
    pub name: String,
}

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to unmount the snapshot
pub struct Unmount {
    /// Filesystem name
    pub fsname: String,
    /// Name of the snapshot
    pub name: String,
}
