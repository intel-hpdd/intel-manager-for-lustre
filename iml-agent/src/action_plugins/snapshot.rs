// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use iml_cmd::{CheckedCommandExt, Command};

pub async fn mount(filesystem_name: String, snapshot_name: String) -> Result<(), ImlAgentError> {
    Command::new("lctl")
        .arg("snapshot_mount")
        .arg("--fsname")
        .arg(filesystem_name)
        .arg("--name")
        .arg(snapshot_name)
        .checked_status()
        .await
        .map_err(|e| e.into())
}

pub async fn unmount(snapshot_name: String) -> Result<(), ImlAgentError> {
    Command::new("lctl")
        .arg("snapshot_unmount")
        .arg("--name")
        .arg(snapshot_name)
        .checked_status()
        .await
        .map_err(|e| e.into())
}
