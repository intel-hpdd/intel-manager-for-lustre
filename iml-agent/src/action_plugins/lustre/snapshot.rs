// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{agent_error::ImlAgentError, device_scanner_client, lustre::lctl};
use chrono::{DateTime, TimeZone, Utc};
use futures::future::try_join_all;
use iml_wire_types::snapshot::{Create, Destroy, List, Mount, Snapshot, Unmount};

type DeviceSnapshots = (String, DateTime<Utc>, DateTime<Utc>, Option<String>);

pub async fn list(l: List) -> Result<Vec<Snapshot>, ImlAgentError> {
    let mut args = vec!["--device", &l.target, "snapshot", "-l"];
    if let Some(name) = &l.name {
        args.push(name);
    }
    let stdout = lctl(args).await?;
    let stdout = stdout.trim();

    if stdout.is_empty() {
        return Ok(vec![]);
    }

    let mounts = device_scanner_client::get_snapshot_mounts().await?;

    let xs = parse_device_snapshots(stdout).into_iter().map(
        |(snapshot_name, create_time, modify_time, comment)| {
            build_snapshot(
                &l.target,
                snapshot_name,
                create_time,
                modify_time,
                comment,
                &mounts,
            )
        },
    );

    let snapshots = try_join_all(xs)
        .await?
        .into_iter()
        .filter_map(|x| x)
        .collect::<Vec<Snapshot>>();

    Ok(snapshots)
}

pub async fn create(c: Create) -> Result<(), ImlAgentError> {
    let use_barrier = match c.use_barrier {
        true => "on",
        false => "off",
    };

    let mut args = vec![
        "snapshot_create",
        "--fsname",
        &c.fsname,
        "--name",
        &c.name,
        "--barrier",
        use_barrier,
    ];

    if let Some(cmnt) = &c.comment {
        args.push("--comment");
        args.push(cmnt);
    }

    lctl(args).await.map(drop)
}

pub async fn destroy(d: Destroy) -> Result<(), ImlAgentError> {
    let mut args = vec!["snapshot_destroy", "--fsname", &d.fsname, "--name", &d.name];
    if d.force {
        args.push("--force");
    }
    lctl(args).await.map(drop)
}

pub async fn mount(m: Mount) -> Result<(), ImlAgentError> {
    let args = &["snapshot_mount", "--fsname", &m.fsname, "--name", &m.name];
    lctl(args).await.map(drop)
}

pub async fn unmount(u: Unmount) -> Result<(), ImlAgentError> {
    let args = &["snapshot_umount", "--fsname", &u.fsname, "--name", &u.name];
    lctl(args).await.map(drop)
}

async fn build_snapshot(
    target: &str,
    snapshot_name: String,
    create_time: DateTime<Utc>,
    modify_time: DateTime<Utc>,
    comment: Option<String>,
    mounts: &[device_types::mount::Mount],
) -> Result<Option<Snapshot>, ImlAgentError> {
    let mut fs_name = target.rsplitn(2, '-').nth(1).map(String::from);
    let filesystem_name = if let Some(name) = fs_name.take() {
        name
    } else {
        return Ok(None);
    };

    let snapshot_fsname = get_snapshot_label(&snapshot_name, &target).await?;
    let snapshot_fsname = if let Some(name) = snapshot_fsname {
        name
    } else {
        return Ok(None);
    };

    let mounted: bool = mounts
        .iter()
        .find_map(|x| {
            let s = x.opts.0.split(',').find(|x| x.starts_with("svname="))?;

            let s = s.split('=').nth(1)?.rsplitn(2, '-').nth(1)?;

            if s == snapshot_fsname {
                return Some(true);
            }

            None
        })
        .unwrap_or_default();

    Ok(Some(Snapshot {
        filesystem_name,
        snapshot_name,
        create_time,
        modify_time,
        comment,
        snapshot_fsname,
        mounted,
    }))
}

async fn get_snapshot_label(
    snapshot_name: &str,
    target: &str,
) -> Result<Option<String>, ImlAgentError> {
    let x = lctl(vec![
        "--device",
        target,
        "snapshot",
        "--get_label",
        snapshot_name,
    ])
    .await?;
    let x = x.trim().rsplitn(2, '-').nth(1).map(String::from);

    Ok(x)
}

fn parse_device_snapshots(x: &str) -> Vec<DeviceSnapshots> {
    x.trim()
        .lines()
        .map(|x| x.splitn(4, ' ').collect::<Vec<_>>())
        .filter_map(|xs| {
            let mut it = xs.into_iter();

            let name = it.next()?.to_string();
            let create = chrono::Utc.timestamp(it.next()?.parse().ok()?, 0);
            let update = chrono::Utc.timestamp(it.next()?.parse().ok()?, 0);
            let comment = it.next().map(String::from);

            Some((name, create, update, comment))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_device_list() {
        let fixture = r#"
testsnapa 1604388443 1604388443
abcd 1604331109 1604331109
testabc 1604331284 1604331284
snaptwo 1604469817 1604469817
testabc4 1604332160 1604332160
1-es01a-2020-11-02T16:10:24Z 1604333425 1604333425 automatically created by IML
1-es01a-2020-11-02T16:40:24Z 1604335225 1604335225 automatically created by IML
snapone 1604469809 1604469809
testsnapz 1604468838 1604468838
1-es01a-2020-11-02T16:20:24Z 1604334025 1604334025 automatically created by IML
1-es01a-2020-11-02T16:30:24Z 1604334625 1604334625 automatically created by IML
1-es01a-2020-11-02T16:00:24Z 1604332825 1604332825 automatically created by IML
snapthree 1604470070 1604470070
testsnapd 1604469760 1604469760
tstsnap 1604335880 1604335880
testsnapb 1604416099 1604416099
1-es01a-2020-11-02T15:50:24Z 1604332227 1604332227 automatically created by IML
testabc3 1604331898 1604331898
mysnap 1604330923 1604330922 mynew snapshot
Test3 1604416497 1604416497
testsnap 1604330047 1604075186
        "#;
        let fixture = fixture;

        let xs = parse_device_snapshots(fixture);

        insta::assert_debug_snapshot!(xs);
    }
}
