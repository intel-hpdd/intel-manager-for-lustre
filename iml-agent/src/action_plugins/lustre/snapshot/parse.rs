// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use chrono::offset::Utc;
use chrono::DateTime;
use combine::{
    attempt, choice, eof,
    error::{ParseError, StreamError},
    many1,
    parser::{char, repeat::take_until},
    satisfy, skip_many,
    stream::{Stream, StreamErrorFor},
    Parser,
};
use iml_wire_types::snapshot::{Detail, Snapshot, Status};
use std::collections::HashMap;

#[derive(PartialEq, Debug)]
enum Segment {
    Snapshot(Snapshot),
    Detail(Detail),
}

fn spaces<I>() -> impl Parser<I, Output = ()>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    skip_many(satisfy(|c| c == ' ' || c == '\t'))
}

fn eol<I>() -> impl Parser<I, Output = ()>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    spaces().skip(choice((char::newline().map(|_| ()), eof())))
}

fn delimiter<I>() -> impl Parser<I, Output = ()>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    spaces().skip(char::char(':').skip(spaces()))
}

fn entry<I>() -> impl Parser<I, Output = (String, String)>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    (
        spaces()
            .with(take_until(attempt(delimiter().or(eol()))))
            .skip(delimiter()),
        take_until(attempt(eol())).skip(eol()),
    )
}

fn parse_date(mut s: String) -> Result<DateTime<Utc>, &'static str> {
    // Can't parse without a timezone.
    // lctl does not print timezone, but uses gettimeofday()
    // to get the time and ctime() to print it.
    s.push_str(" +00:00");
    DateTime::parse_from_str(&s, "%a %b %e %T %Y %#z")
        .map(|t| t.with_timezone(&Utc))
        .map_err(|_| "date string like 'Fri Aug  7 16:43:06 2020'")
}

fn mk_detail(mut m: HashMap<String, String>) -> Result<Detail, &'static str> {
    let snapshot_fsname = m.remove("snapshot_fsname").ok_or("snapshot_fsname")?;
    let modify_time_s = m.remove("modify_time").ok_or("modify_time")?;
    let create_time_s = m.remove("create_time").ok_or("create_time")?;
    let status_s = m.remove("status").ok_or("status")?;
    let comment = m.remove("comment");
    let snapshot_role = m.remove("snapshot_role");

    let modify_time = parse_date(modify_time_s)?;
    let create_time = parse_date(create_time_s)?;
    let status = match status_s.as_str() {
        "mounted" => Some(Status::Mounted),
        "not mount" => Some(Status::NotMounted),
        _ => None,
    };
    Ok(Detail {
        snapshot_role,
        snapshot_fsname,
        modify_time,
        create_time,
        status,
        comment,
    })
}

fn mk_snapshot(mut m: HashMap<String, String>) -> Result<Snapshot, &'static str> {
    let filesystem_name = m.remove("filesystem_name").ok_or("filesystem_name")?;
    let snapshot_name = m.remove("snapshot_name").ok_or("snapshot_name")?;
    let details = if m.is_empty() {
        vec![]
    } else {
        vec![mk_detail(m)?]
    };
    Ok(Snapshot {
        filesystem_name,
        snapshot_name,
        details,
    })
}

fn map<I>() -> impl Parser<I, Output = Segment>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    char::spaces()
        .with(many1(attempt(entry())))
        .skip(char::spaces())
        .and_then(|m: HashMap<String, String>| {
            if m.get("filesystem_name").is_some() {
                mk_snapshot(m)
                    .map(Segment::Snapshot)
                    .map_err(StreamErrorFor::<I>::expected_static_message)
            } else if m.get("snapshot_role").is_some() {
                mk_detail(m)
                    .map(Segment::Detail)
                    .map_err(StreamErrorFor::<I>::expected_static_message)
            } else {
                Err(StreamErrorFor::<I>::expected_static_message(
                    "snapshot or snapshot detail",
                ))
            }
        })
}

pub(crate) fn parse<I>() -> impl Parser<I, Output = Vec<Snapshot>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(map()).and_then(|ss: Vec<Segment>| {
        let mut it = ss.iter();
        let s = it.next().unwrap();
        if let Segment::Snapshot(snap) = s {
            let (s, mut v) = it.fold((snap.clone(), vec![]), |(mut snap, mut v), s| {
                match s {
                    Segment::Snapshot(sn) => {
                        v.push(snap);
                        snap = sn.clone();
                    }
                    Segment::Detail(d) => {
                        snap.details.push(d.clone());
                    }
                }
                (snap, v)
            });
            v.push(s);
            Ok(v)
        } else {
            Err(StreamErrorFor::<I>::expected_static_message("snapshot"))
        }
    })
}

#[cfg(test)]
mod test {
    use chrono::offset::TimeZone;
    use chrono::offset::Utc;
    use combine::EasyParser;
    use iml_wire_types::snapshot::{Detail, Snapshot, Status};

    #[test]
    fn parse_date() {
        let r = super::parse_date("Fri Aug  7 16:43:06 2020".to_string());
        assert_eq!(r, Ok(Utc.ymd(2020, 8, 7).and_hms(16, 43, 06)));
    }

    #[test]
    fn parse_simple() {
        // white space is intentional:
        let sample = r#"

filesystem_name: zfsmo
    snapshot_name  : bar
    snapshot_fsname: 16c3a547
    create_time: Fri Aug  7 16:43:06 2020
modify_time: Fri Aug  7 16:43:06 2020
    status: not mount



        filesystem_name: zfsmo
        snapshot_name: foo
        create_time: Fri Aug  7 16:29:30 2020
        modify_time: Fri Aug  7 17:29:30 2020
        snapshot_fsname: 6f27d503
        comment: hello world
        status: mounted

        "#;

        let snaps = vec![
            Snapshot {
                filesystem_name: "zfsmo".into(),
                snapshot_name: "bar".into(),
                details: vec![Detail {
                    comment: None,
                    create_time: Utc.ymd(2020, 8, 7).and_hms(16, 43, 06),
                    snapshot_fsname: "16c3a547".into(),
                    modify_time: Utc.ymd(2020, 8, 7).and_hms(16, 43, 06),
                    snapshot_role: None,
                    status: Some(Status::NotMounted),
                }],
            },
            Snapshot {
                filesystem_name: "zfsmo".into(),
                snapshot_name: "foo".into(),
                details: vec![Detail {
                    comment: Some("hello world".into()),
                    create_time: Utc.ymd(2020, 8, 7).and_hms(16, 29, 30),
                    snapshot_fsname: "6f27d503".into(),
                    modify_time: Utc.ymd(2020, 8, 7).and_hms(17, 29, 30),
                    snapshot_role: None,
                    status: Some(Status::Mounted),
                }],
            },
        ];

        assert_eq!(super::parse().easy_parse(sample), Ok((snaps, "")));
    }

    #[test]
    fn parse_detailed() {
        let sample = r#"
filesystem_name: zfsmo
snapshot_name: bar

snapshot_role: MGS
modify_time: Fri Aug  7 16:43:06 2020
create_time: Fri Aug  7 16:43:06 2020
snapshot_fsname: 16c3a547
status: not mount

filesystem_name: zfsmo
snapshot_name: foo

snapshot_role: MGS
comment: hello world
modify_time: Fri Aug  7 16:29:30 2020
create_time: Fri Aug  7 16:29:30 2020
snapshot_fsname: 6f27d503
status: not mount

snapshot_role: MDT0000
create_time: Fri Aug  7 16:29:30 2020
modify_time: Fri Aug  7 16:29:30 2020
snapshot_fsname: 6f27d503
comment: hello world
status: not mount
        "#;

        let snaps = vec![
            Snapshot {
                filesystem_name: "zfsmo".into(),
                snapshot_name: "bar".into(),
                details: vec![Detail {
                    comment: None,
                    create_time: Utc.ymd(2020, 8, 7).and_hms(16, 43, 06),
                    snapshot_fsname: "16c3a547".into(),
                    modify_time: Utc.ymd(2020, 8, 7).and_hms(16, 43, 06),
                    snapshot_role: Some("MGS".into()),
                    status: Some(Status::NotMounted),
                }],
            },
            Snapshot {
                filesystem_name: "zfsmo".into(),
                snapshot_name: "foo".into(),
                details: vec![
                    Detail {
                        comment: Some("hello world".into()),
                        create_time: Utc.ymd(2020, 8, 7).and_hms(16, 29, 30),
                        snapshot_fsname: "6f27d503".into(),
                        modify_time: Utc.ymd(2020, 8, 7).and_hms(16, 29, 30),
                        snapshot_role: Some("MGS".into()),
                        status: Some(Status::NotMounted),
                    },
                    Detail {
                        comment: Some("hello world".into()),
                        create_time: Utc.ymd(2020, 8, 7).and_hms(16, 29, 30),
                        snapshot_fsname: "6f27d503".into(),
                        modify_time: Utc.ymd(2020, 8, 7).and_hms(16, 29, 30),
                        snapshot_role: Some("MDT0000".into()),
                        status: Some(Status::NotMounted),
                    },
                ],
            },
        ];

        assert_eq!(super::parse().easy_parse(sample), Ok((snaps, "")));
    }
}
