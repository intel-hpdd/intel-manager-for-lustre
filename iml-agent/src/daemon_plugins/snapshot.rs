// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

//! Snapshot listing plugin
//!
//! This module is responsible for continually discovering snapshots of filesystems.
//!
//!

use crate::{
    action_plugins::lustre::snapshot,
    agent_error::ImlAgentError,
    daemon_plugins::{DaemonPlugin, Output},
    device_scanner_client,
    lustre::list_mdt0s,
};
use async_trait::async_trait;
use futures::{
    future::{try_join_all, AbortHandle, Abortable},
    lock::Mutex,
    Future, FutureExt, TryFutureExt,
};
use iml_wire_types::snapshot::{List, Snapshot};
use std::{
    collections::{HashMap, HashSet},
    pin::Pin,
    sync::Arc,
    time::Duration,
};
use tokio::time::delay_for;

struct State {
    updated: bool,
    output: Output,
}

#[derive(Debug)]
pub(crate) struct SnapshotList {
    reader: Option<AbortHandle>,
    state: Arc<Mutex<State>>,
}

pub(crate) fn create() -> SnapshotList {
    SnapshotList {
        reader: None,
        state: Arc::new(Mutex::new(State {
            updated: false,
            output: None,
        })),
    }
}

async fn list() -> Result<Option<HashMap<String, Vec<Snapshot>>>, ImlAgentError> {
    // list the mounts
    // remove any mdt0's that are snapshots
    let snapshot_mounts = device_scanner_client::get_snapshot_mounts()
        .await?
        .into_iter()
        .filter_map(|x| {
            let s = x.opts.0.split(',').find(|x| x.starts_with("svname="))?;

            let s = s.split('=').nth(1)?;

            Some(s.to_string())
        })
        .collect::<HashSet<String>>();

    tracing::debug!("snapshot mounts {:?}", snapshot_mounts);

    let xs = list_mdt0s().await.into_iter().collect::<HashSet<String>>();
    let xs = xs
        .difference(&snapshot_mounts)
        .map(|x| x.to_string())
        .collect::<Vec<String>>();

    tracing::debug!("mdt0s: {:?}", &xs);

    if xs.is_empty() {
        tracing::debug!("No mdt0's. Returning `None`.");

        return Ok(None);
    }

    let xs = xs.into_iter().map(|x| async move {
        let fs_name = x.rsplitn(2, '-').nth(1);
        tracing::debug!("fs_name is: {:?}", fs_name);

        if let Some(fs_name) = fs_name {
            snapshot::list(List {
                target: x.to_string(),
                name: None,
            })
            .inspect_err(|x| {
                tracing::debug!("Error calling snapshot list: {:?}", x);
            })
            .await
            .map(|x| (fs_name.to_string(), x))
        } else {
            tracing::debug!("No fs_name. Returning MarkerNotFound error.");

            Err(ImlAgentError::MarkerNotFound)
        }
    });

    let snapshots: HashMap<String, Vec<Snapshot>> = try_join_all(xs).await?.into_iter().collect();

    tracing::debug!("snapshots: {:?}", snapshots);

    Ok(Some(snapshots))
}

async fn try_update_state(state: &Arc<Mutex<State>>) -> Result<(), ImlAgentError> {
    let snapshots = list().await?;

    let output = snapshots.map(|xs| serde_json::to_value(xs)).transpose()?;

    let mut lock = state.lock().await;

    if lock.output != output {
        tracing::debug!("Snapshot output changed. Updating.");

        lock.output = output;
        lock.updated = true;
    }

    Ok(())
}

#[async_trait]
impl DaemonPlugin for SnapshotList {
    fn start_session(
        &mut self,
    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
        let state = Arc::clone(&self.state);

        let (reader, reader_reg) = AbortHandle::new_pair();
        self.reader = Some(reader);

        async move {
            tokio::spawn(Abortable::new(
                async move {
                    loop {
                        if let Err(e) = try_update_state(&state).await {
                            tracing::debug!(
                                "Error calling list(): {:?}. Not processing snapshots.",
                                e
                            );
                        }

                        delay_for(Duration::from_secs(10)).await;
                    }
                },
                reader_reg,
            ));

            Ok(None)
        }
        .boxed()
    }

    fn update_session(
        &self,
    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
        let state = Arc::clone(&self.state);
        async move {
            let mut lock = state.lock().await;

            if lock.updated {
                lock.updated = false;
                Ok(lock.output.clone())
            } else {
                Ok(None)
            }
        }
        .boxed()
    }

    async fn teardown(&mut self) -> Result<(), ImlAgentError> {
        if let Some(reader) = &self.reader {
            reader.abort();
            self.reader = None;
        }

        Ok(())
    }
}
