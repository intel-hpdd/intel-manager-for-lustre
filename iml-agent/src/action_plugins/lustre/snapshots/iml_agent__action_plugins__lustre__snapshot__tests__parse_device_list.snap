---
source: iml-agent/src/action_plugins/lustre/snapshot.rs
expression: xs
---
[
    (
        "testsnapa",
        2020-11-03T07:27:23Z,
        2020-11-03T07:27:23Z,
        None,
    ),
    (
        "abcd",
        2020-11-02T15:31:49Z,
        2020-11-02T15:31:49Z,
        None,
    ),
    (
        "testabc",
        2020-11-02T15:34:44Z,
        2020-11-02T15:34:44Z,
        None,
    ),
    (
        "snaptwo",
        2020-11-04T06:03:37Z,
        2020-11-04T06:03:37Z,
        None,
    ),
    (
        "testabc4",
        2020-11-02T15:49:20Z,
        2020-11-02T15:49:20Z,
        None,
    ),
    (
        "1-es01a-2020-11-02T16:10:24Z",
        2020-11-02T16:10:25Z,
        2020-11-02T16:10:25Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "1-es01a-2020-11-02T16:40:24Z",
        2020-11-02T16:40:25Z,
        2020-11-02T16:40:25Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "snapone",
        2020-11-04T06:03:29Z,
        2020-11-04T06:03:29Z,
        None,
    ),
    (
        "testsnapz",
        2020-11-04T05:47:18Z,
        2020-11-04T05:47:18Z,
        None,
    ),
    (
        "1-es01a-2020-11-02T16:20:24Z",
        2020-11-02T16:20:25Z,
        2020-11-02T16:20:25Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "1-es01a-2020-11-02T16:30:24Z",
        2020-11-02T16:30:25Z,
        2020-11-02T16:30:25Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "1-es01a-2020-11-02T16:00:24Z",
        2020-11-02T16:00:25Z,
        2020-11-02T16:00:25Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "snapthree",
        2020-11-04T06:07:50Z,
        2020-11-04T06:07:50Z,
        None,
    ),
    (
        "testsnapd",
        2020-11-04T06:02:40Z,
        2020-11-04T06:02:40Z,
        None,
    ),
    (
        "tstsnap",
        2020-11-02T16:51:20Z,
        2020-11-02T16:51:20Z,
        None,
    ),
    (
        "testsnapb",
        2020-11-03T15:08:19Z,
        2020-11-03T15:08:19Z,
        None,
    ),
    (
        "1-es01a-2020-11-02T15:50:24Z",
        2020-11-02T15:50:27Z,
        2020-11-02T15:50:27Z,
        Some(
            "automatically created by IML",
        ),
    ),
    (
        "testabc3",
        2020-11-02T15:44:58Z,
        2020-11-02T15:44:58Z,
        None,
    ),
    (
        "mysnap",
        2020-11-02T15:28:43Z,
        2020-11-02T15:28:42Z,
        Some(
            "mynew snapshot",
        ),
    ),
    (
        "Test3",
        2020-11-03T15:14:57Z,
        2020-11-03T15:14:57Z,
        None,
    ),
    (
        "testsnap",
        2020-11-02T15:14:07Z,
        2020-10-30T16:26:26Z,
        None,
    ),
]
